File: gparsers/src/main/kotlin/org/parser/CFPQCsvGraph.kt
package org.parser

import org.neo4j.graphdb.GraphDatabaseService
import org.parser.combinators.*
import org.parser.neo4j.DefaultNeo4jCombinators.inE
import org.parser.neo4j.DefaultNeo4jCombinators.inV
import org.parser.neo4j.DefaultNeo4jCombinators.outE
import org.parser.neo4j.DefaultNeo4jCombinators.outV
import org.parser.neo4j.DefaultNeo4jCombinators.throughInE
import org.parser.neo4j.DefaultNeo4jCombinators.throughOutE
import org.parser.neo4j.DefaultNeo4jGraph
import org.parser.neo4j.DefaultNeo4jVertexState
import org.parser.shared.loadCsvEdges
import java.io.InputStream

object CFPQCsvGraph {
    private fun getFileStream(file: String): InputStream {
        return this.javaClass.getResourceAsStream("/cfpq/$file")
            ?: throw RuntimeException("Can't find cfpq resource $file")
    }

    fun getGraph(file: String, db: GraphDatabaseService): DefaultNeo4jGraph {
        val edges = loadCsvEdges(getFileStream(file)).toList()
        val nodesCount = edges.flatMap { listOf(it.from, it.to) }.toSet().size
        return GParsers.edgesToNeo4jGraph(db, edges, nodesCount)
    }

    fun firstGrammar(): BaseParser<DefaultNeo4jVertexState, DefaultNeo4jVertexState, Any> {
        val subclassof1 = throughInE { it.label == "subClassOf" }
        val subclassof = throughOutE { it.label == "subClassOf" }
        val type1 = throughInE { it.label == "type" }
        val type = throughOutE { it.label == "type" }
        val S = LazyParser<DefaultNeo4jVertexState, DefaultNeo4jVertexState, Any>()
        S.p =
                (subclassof1 seq (S or eps()) seq subclassof) or
                (type1 seq (S or eps()) seq type)

        return S
    }
}

File: meerkat/src/main/scala/org/parser/CFPQCsvGraph.scala
package org.parser

import org.meerkat.Syntax.syn
import org.meerkat.graph.neo4j.Neo4jInput
import org.meerkat.graph.neo4j.Neo4jInput.Entity
import org.meerkat.parsers.Parsers.{inE, outE, ε}
import org.meerkat.parsers.Parsers.{Symbol, V, inE, outE, ε}
import org.neo4j.graphdb.GraphDatabaseService
import org.parser.Meerkat.{L, N}
import org.parser.shared.CsvLoaderKt
import org.meerkat.parsers.AbstractCPSParsers
import org.meerkat.Syntax.syn
import org.meerkat.graph.neo4j.Neo4jInput
import org.meerkat.graph.neo4j.Neo4jInput.Entity
import org.meerkat.graph.parseGraphFromAllPositions
import org.meerkat.input.{GraphxInput, Input}
import org.meerkat.parsers.{AbstractCPSParsers, Parsers}
import org.meerkat.parsers.Parsers.{Symbol, V, inE, outE, ε}
import org.neo4j.test.TestGraphDatabaseFactory
import org.parser.shared.Edge

import scala.collection.JavaConverters.asScalaBufferConverter

import java.io.InputStream
import scala.collection.convert.ImplicitConversions.`iterable AsScalaIterable`

object CFPQCsvGraph {
  private def getFileStream(file: String): InputStream = {
    getClass.getResourceAsStream(f"/cfpq/$file")
  }

  def getGraph(file: String): Neo4jInput = {
    val edges = CsvLoaderKt.loadCsvEdges(getFileStream(file)).toList
    val nodesCount = edges.flatMap { case it => List(it.getFrom, it.getTo) }.toSet.size
    Meerkat.edgesToNeo4jGraph(file, edges, nodesCount)
  }


  val subclassof1: Symbol[L, N, _] = inE("subClassOf")
  val subclassof: Symbol[L, N, _] = outE("subClassOf")
  val type1: Symbol[L, N, _] = inE("type")
  val _type: Symbol[L, N, _] = outE("type")

  val grammar: Symbol[L, N, _] = syn((subclassof1 ~ syn(grammar | ε) ~ subclassof) | (type1 ~ syn(grammar | ε) ~ _type))


}
